### PR TITLE
add support for global options handler

### PR DESCRIPTION
**Pull request by ori-shalom**
Allow users to define options handler to respond to any `OPTIONS` request that isn't explicitly defined as an operation in the spec.

The default behavior is to respond with 204 No content and the [`Allow` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Allow).

Users can use this handler to respond to CORS preflight requests by setting a custom OptionsHandler and add [`Access-Control-Allow-Methods` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods) that return the methods listed in the `Allow` header.

The implementation is based on how [httprouter](https://github.com/julienschmidt/httprouter?tab=readme-ov-file#automatic-options-responses-and-cors) allow to handle OPTIONS requests.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,24 +6,27 @@ retract v1.0.0 // Published accidentally.
 
 require (
 	github.com/getkin/kin-openapi v0.117.0
-	github.com/google/uuid v1.3.0
-	github.com/invopop/jsonschema v0.4.0
+	github.com/google/uuid v1.4.0
```
Comment (nirhaas) [inline go.mod:9]: Can you `go mod tidy` this file?

File: router/options.go [hunk 2]
```diff
@@ -126,6 +127,13 @@ type Options struct {
 	// One use for this option is when your spec defines the entire API of your app but the implementation is spread to multiple microservices.
 	// With this option you can define list of operations that are to be implemented by other microservices.
 	ExcludeOperations []string
+
+	// OptionsHandler defines a handler that is called for every OPTIONS request of any path that is defined in the spec.
+	// This handler allows you to define a response for the OPTIONS preflight request and control the allowed methods or set CORS headers.
+	// The handler will receive the allowed methods as a slice of strings computed by the router based on the spec.
+	// If not set the router will default to a handler that will set the allowed methods as an `Allow` header.
+	// Set to nil to disable the automatic handling of OPTIONS requests.
```
Comment (nirhaas) [inline router/options.go:135]: What is the difference between the two options? unclear

File: router/options.go [hunk 3]
```diff
@@ -126,6 +127,13 @@ type Options struct {
 	// One use for this option is when your spec defines the entire API of your app but the implementation is spread to multiple microservices.
 	// With this option you can define list of operations that are to be implemented by other microservices.
 	ExcludeOperations []string
+
+	// OptionsHandler defines a handler that is called for every OPTIONS request of any path that is defined in the spec.
+	// This handler allows you to define a response for the OPTIONS preflight request and control the allowed methods or set CORS headers.
+	// The handler will receive the allowed methods as a slice of strings computed by the router based on the spec.
```
Comment (nirhaas) [inline router/options.go:133]: What does this line mean?

File: router/core.go [hunk 4]
```diff
@@ -150,3 +172,27 @@ func defaultRecoverBehaviour(writer http.ResponseWriter) {
...
+func notFoundHandler(oa *openapi) http.HandlerFunc {
+	if oa.options.OptionsHandler == nil {
+		return func(writer http.ResponseWriter, request *http.Request) {
+			writer.WriteHeader(http.StatusNotFound)
+		}
+	}
+
+	return func(writer http.ResponseWriter, request *http.Request) {
+		if request.Method != http.MethodOptions {
+			writer.WriteHeader(http.StatusNotFound)
```
Comment (nirhaas) [inline router/core.go:190]: How can this happen?

File: router/core.go [hunk 5]
```diff
@@ -150,3 +172,27 @@ func defaultRecoverBehaviour(writer http.ResponseWriter) {
...
+			return
+		}
+
+		// If we get to NotFound it means that the path does not exist so we can set the "Allow" header to "OPTIONS" only and call the OPTIONS handler.
+		writer.Header().Set("Allow", "OPTIONS")
```
Comment (nirhaas) [inline router/core.go:195]: Make sure it returns 404 plssss

File: router/core.go [hunk 6]
```diff
@@ -54,6 +56,26 @@ func createMainRouterHandler(oa *openapi) (http.Handler, error) {
...
+
+	if !router.HandleOPTIONS {
+		return
+	}
+
+	// GlobalOPTIONS is used for the OPTIONS handler for all paths.
+	// It receives all registered methods for a path in the "Allow" header.
+	router.GlobalOPTIONS = http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+		oa.options.OptionsHandler.ServeHTTP(writer, request)
+	})
```
Comment (nirhaas) [inline router/core.go:76]:
```suggestion
	router.GlobalOPTIONS = oa.options.OptionsHandler
```

File: router/core.go [hunk 7]
```diff
@@ -54,6 +56,26 @@ func createMainRouterHandler(oa *openapi) (http.Handler, error) {
...
+// setOptionsHandlers sets the OPTIONS handlers for the router.
+func setOptionsHandlers(router *httprouter.Router, oa *openapi) {
+	router.HandleOPTIONS = oa.options.OptionsHandler != nil
+
+	// If HandleOptions is nil, NotFound simply return 404 for any request.
+	// When HandleOptions is not nil, NotFound for any OPTIONS request will call the global OPTIONS handler to return the "Allow: OPTIONS" header.
+	// For any other request it will return 404
+	router.NotFound = notFoundHandler(oa)
+
+	if !router.HandleOPTIONS {
```
Comment (nirhaas) [inline router/core.go:68]: Can be removed

File: go.mod (hunk 1 above)
Comment (ori-shalom) [reply go.mod:9]: I did. This is the result

**Review by nirhaas — APPROVED** [review]

**Pull request merged**